### PR TITLE
cask: add audit for incorrect signing

### PR DESCRIPTION
**Pull request by SMillerDev**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----
Since often upstreams suggest people disable quarantine instead of fixing their broken signatures, we should probably check if the signature is valid in the audit.

### CODE REVIEW

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by SMillerDev** [comment]
Changed it to a separate flag.

File: Library/Homebrew/cask/audit.rb [hunk 1]
```diff
@@ -550,6 +554,33 @@ def check_download
       add_error "download not possible: #{e}"
     end
 
+    def check_signing
+      return if !signing? || cask.url.blank?
+
+      odebug "Auditing signing"
+      artifacts = cask.artifacts.filter { |k| k.is_a?(Artifact::Pkg) || k.is_a?(Artifact::App) }
+
+      puts artifacts
```
Comment (Rylan12) [inline Library/Homebrew/cask/audit.rb:563]:
```suggestion
```

I assume this is just a debug message, but if not, I don't think we want anything to be printed unless there's an issue

**Review by MikeMcQuaid — COMMENTED** [review]
Looking good so far. When CI is green would like to see some of the new code touched by at least some unit tests.

**Comment by SMillerDev** [comment]
Only the actual signing command is left out for now since that depends on a specific macOS version for output checking. Otherwise I think it is good to go.

File: Library/Homebrew/cask/audit.rb [hunk 2]
```diff
@@ -20,28 +20,31 @@ class Audit
...
-    attr_predicate :appcast?, :new_cask?, :strict?, :online?, :token_conflicts?
+    attr_predicate :appcast?, :new_cask?, :strict?, :signing?, :online?, :token_conflicts?
 
     def initialize(cask, appcast: nil, download: nil, quarantine: nil,
-                   token_conflicts: nil, online: nil, strict: nil,
+                   token_conflicts: nil, online: nil, strict: nil, signing: nil,
                    new_cask: nil)
 
-      # `new_cask` implies `online` and `strict`
+      # `new_cask` implies `online`, `token_conflicts`, `strict` and signing
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/audit.rb:29]:
```suggestion
      # `new_cask` implies `online`, `token_conflicts`, `strict` and `signing`
```

**Review by MikeMcQuaid — APPROVED** [review]
One non-blocking comment but otherwise 👍🏻

**Comment by bevanjkay** [comment]
@SMillerDev Looks like this commit is causing a couple of false positive errors - https://github.com/Homebrew/homebrew-cask/runs/7859932752?check_suite_focus=true
https://github.com/Homebrew/homebrew-cask-versions/runs/7859526329?check_suite_focus=true

I don't have time to fully check it out (it's late here), but just flagging this PR as I've traced the errors back to here.

**Comment by SMillerDev** [comment]
Okay, somehow my artifact filtering isn't working. I would have expected this line to output a list of only eligible artifacts.
```rb
artifacts = cask.artifacts.select { |k| k.is_a?(Artifact::Pkg) || k.is_a?(Artifact::App) }
```

File: Library/Homebrew/cask/audit.rb [hunk 3]
```diff
@@ -550,6 +554,33 @@ def check_download
...
+
+      downloaded_path = download.fetch
+      primary_container = UnpackStrategy.detect(downloaded_path, type: @cask.container&.type, merge_xattrs: true)
+
+      return if primary_container.nil?
+
+      Dir.mktmpdir do |tmpdir|
+        tmpdir = Pathname(tmpdir)
+        primary_container.extract_nestedly(to: tmpdir, basename: downloaded_path.basename, verbose: false)
+        cask.artifacts.each do |artifact|
```
Comment (Rylan12) [inline Library/Homebrew/cask/audit.rb:574]: I think this is the problem. We want `artifacts` not `cask.artifacts`, right?

Comment (SMillerDev) [reply Library/Homebrew/cask/audit.rb:574]: I think you're right here 😅

**Pull request merged**